### DISCUSSION TITLE
[express-serve-static-core] Route param inference does not handle `mergeParams`

### DISCUSSION THREAD

**Discussion by thw0rted** [DefinitelyTyped/DefinitelyTyped#64437] [Issues with a @types package]
When a `Router` is constructed with the option `mergeParams`, it merges the `req.params` map with its parent Router (or Application, etc).  This means that the type of the first argument to the callback on e.g. `Router({mergeParams: true}).get('/foo/:fid')` cannot be inferred at compile-time, since we don't know what parent the Router will be attached to -- attaching this router to a parent with `app.use('/bar/:bid', myRouter)` will supply a `bid` param for all routes.  This leads to [problems like this one](https://stackoverflow.com/q/72395842/26286), where the type system tells us that parameters we know should exist are not defined, and manual casts are required.

The safest thing to do here would be to return a more broadly typed interface from `Router` when `mergeParams` is true, then fall back to `Record<string,string>` for the type of `req.Params`.  I think there might also be value in providing an explicit generic param so I could write `Router<'param1'|'param2'>({mergeParams: true}).post(...)` (or `Router<['param1', 'param2']>`, whichever makes more sense) and the top level param keys would be included in the generic value passed in to `Request` in the callback.

[a comment by an automated account is not shown]

**Comment by sgtdck** [comment]
As a workaround you can use generics:

```ts
const router = Router({mergeParams: true})
router.get<'/foo/:id', { fid: string, bid: string }>('/foo/:fid', async (req) => {
  const { fid, bid } = req.params;
})
```

It's a little convoluted, and having this inferred from parent would be ideal, but I don't think this is easily achieved.

Still, it's better to be explicit than typecasting.

Edit: I wrote up a more extensive way of doing it here https://stackoverflow.com/a/77366876/1402826
